python/ray/data/_internal/execution/operators/zip_operator.py

```diff
@@ -236,11 +236,28 @@ def _zip_one_block(
         # Swap blocks if ordering was inverted during block alignment splitting.
         block, other_block = other_block, block
     # Zip block and other blocks.
-    result = BlockAccessor.for_block(block).zip(other_block)
+    result = _zip_blocks_with_type_check(block, other_block)
     br = BlockAccessor.for_block(result)
     return result, br.get_metadata(input_files=[], exec_stats=stats.build())
 
 
 def _get_num_rows_and_bytes(block: Block) -> Tuple[int, int]:
     block = BlockAccessor.for_block(block)
     return block.num_rows(), block.size_bytes()
+
+
+def _zip_blocks_with_type_check(block: Block, other_block: Block) -> Block:
+    """Check that block types are the same before zipping. If none of the block
+    are pyarrow blocks, raise ValueError.
+    """
+    if not isinstance(block, type(other_block)):
+        from pyarrow import Table
+
+        if not any(isinstance(blk, Table) for blk in (block, other_block)):
+            raise ValueError(
+                "Cannot zip {} with block of type {}".format(
+                    type(block), type(other_block)
+                )
+            )
```

Comment:
We can later opt to normalize these blocks instead of raising `ValueError` like https://github.com/ray-project/ray/pull/39960

Reply:
agreed, i'm thinking we can call `normalize_blocks([block, other_block)` to ensure both blocks are the same type, then we can simply maintain the current `result = BlockAccessor.for_block(block).zip(other_block)`.